contracts/0.8.9/lib/WithdrawalRequests.sol

```diff
@@ -0,0 +1,122 @@
+
+    error WithdrawalRequestFeeReadFailed();
+    error InvalidPubkeyLength(bytes pubkey);
+    error WithdrawalRequestAdditionFailed(bytes pubkey, uint256 amount);
+    error NoWithdrawalRequests();
+    error PartialWithdrawalRequired(bytes pubkey);
+
+    event WithdrawalRequestAdded(bytes pubkey, uint256 amount);
+
+    /**
+     * @dev Adds full withdrawal requests for the provided public keys.
+     *      The validator will fully withdraw and exit its duties as a validator.
+     * @param pubkeys An array of public keys for the validators requesting full withdrawals.
+     */
+    function addFullWithdrawalRequests(
+        bytes[] calldata pubkeys
+    ) internal {
+        uint256 keysCount = pubkeys.length;
+        uint64[] memory amounts = new uint64[](keysCount);
+
+        _addWithdrawalRequests(pubkeys, amounts);
+    }
+
+    /**
+     * @dev Adds partial withdrawal requests for the provided public keys with corresponding amounts.
+     *      A partial withdrawal is any withdrawal where the amount is greater than zero.
+     *      This allows withdrawal of any balance exceeding 32 ETH (e.g., if a validator has 35 ETH, up to 3 ETH can be withdrawn).
+     *      However, the protocol enforces a minimum balance of 32 ETH per validator, even if a higher amount is requested.
+     * @param pubkeys An array of public keys for the validators requesting withdrawals.
+     * @param amounts An array of corresponding withdrawal amounts for each public key.
+     */
+    function addPartialWithdrawalRequests(
+        bytes[] calldata pubkeys,
+        uint64[] calldata amounts
+    ) internal {
+        uint256 keysCount = pubkeys.length;
+        if (keysCount != amounts.length) {
+            revert MismatchedArrayLengths(keysCount, amounts.length);
+        }
+
+        uint64[] memory _amounts = new uint64[](keysCount);
+        for (uint256 i = 0; i < keysCount; i++) {
+            if (amounts[i] == 0) {
+                revert PartialWithdrawalRequired(pubkeys[i]);
+            }
+
+            _amounts[i] = amounts[i];
+        }
+
+        _addWithdrawalRequests(pubkeys, _amounts);
+    }
+
+    /**
+     * @dev Retrieves the current withdrawal request fee.
+     * @return The minimum fee required per withdrawal request.
+     */
+    function getWithdrawalRequestFee() internal view returns (uint256) {
+        (bool success, bytes memory feeData) = WITHDRAWAL_REQUEST.staticcall("");
+
+        if (!success) {
+            revert WithdrawalRequestFeeReadFailed();
+        }
+
+        return abi.decode(feeData, (uint256));
+    }
+
+    function _addWithdrawalRequests(
+        bytes[] calldata pubkeys,
+        uint64[] memory amounts
+    ) internal {
+        uint256 keysCount = pubkeys.length;
+        if (keysCount == 0) {
+            revert NoWithdrawalRequests();
+        }
+
+        uint256 minFeePerRequest = getWithdrawalRequestFee();
```

Comment:
Doesn't fee increase with each request?

Reply:
No, the fee will not increase with each request.  Inside the transaction, all requests will have the same fee.

EIP 7002 uses block-by-block behavior. 

If block N processes X requests, then at the end of block N  the number of withdrawal requests that the chain has processed relative to the “targeted” number increases by X - TARGET_WITHDRAWAL_REQUESTS_PER_BLOCK, and so the fee in block N+1 increases by a factor of e**((X - TARGET_WITHDRAWAL_REQUESTS_PER_BLOCK) / WITHDRAWAL_REQUEST_FEE_UPDATE_FRACTION).